lib/src/main/java/graphql/nadel/validation/util/NadelGetReachableTypes.kt

```diff
@@ -35,246 +37,291 @@ import graphql.schema.GraphQLTypeReference
 import graphql.schema.GraphQLUnionType
 import graphql.schema.GraphQLUnmodifiedType
 
+internal sealed class NadelReferencedType {
+    abstract val name: String
+
+    data class OrdinaryType(
+        override val name: String,
+    ) : NadelReferencedType()
+
+    data class VirtualType(
+        override val name: String,
+        val backingType: String,
+    ) : NadelReferencedType()
+}
+
 context(NadelValidationContext)
-internal fun getReachableTypeNames(
+internal fun getReferencedTypeNames(
     service: Service,
     definitionNames: List<String>,
-): Set<String> {
-    // We need a mutable Set so we can keep track of what types we've visited to avoid StackOverflows
-    val reachableTypeNames = mutableSetOf<String>()
-
-    // We keep these so we can add breakpoints to see where a type came from
-    fun add(name: String) = reachableTypeNames.add(name)
-    fun addAll(elements: Collection<String>) = reachableTypeNames.addAll(elements)
-
-    val traverser = object : NadelFastSchemaTraverser.Visitor {
```

Comment:
Sorry most of this is a reformat due to converting this anonyoumous declaration into a proper class.

The formatting changed.